### ISSUE TITLE
Structural pattern match of self over self's class incorrectly reports inexhaustive match

### ISSUE THREAD

**Issue by thisisrandy** [microsoft/pyright#3049]
**Describe the bug**
When using structural pattern matching on `self` in an instance method, pyright reports, 'Cases within match statement do not exhaustively handle all values. Unhandled type: "Self@CLASS_NAME"...'

**To Reproduce**
The most basic case is something like the following:
```python
class Foo:
    def __str__(self) -> str:
        match self:
            case Foo(): return "baz"
```
However, there is little practical reason to do such a thing, so here is the motivating application using `Enum`s:
```python
from enum import Enum, auto


class Foo(Enum):
    bar = auto()

    def __str__(self) -> str:
        match self:
            case Foo.bar:
                return "bar"
```

In either case, pyright reports, 'Unhandled type: "Self@Foo".' Running e.g. `print(f"{Foo.bar}")` prints "bar" as expected.

Two notes:
1. Adding `assert isinstance(self, Foo)` above the `match` statement quiets the error. Annotating `self: Foo` does the same.
2. Seemingly related is the alternate formulation `if self is Foo.bar: return "bar"`, for which pyright complains, "Function with declared type of "str" must return value." Curiously, adding an assertion to that formulation does *not* quiet the error, although annotating `self` does.

**Expected behavior**
The type checker should recognize that `self`, being an instance of `Foo`, can only be `Foo.bar`. It's unclear to me what `Self@CLASS_NAME` as the type denotes (why isn't the type just the class?), but some incongruity between the `Self@` type and the class type seems to be at the heart of the issue.

**VS Code extension or command-line**
Pylance v2022.2.1

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

Your analysis is correct. The type narrowing logic wasn't properly handling the negative (not matched) case for class patterns when the subject expression type was a bound TypeVar. The `Self` type is effectively a TypeVar that is bound to `Foo`.

**Comment by erictraut** [comment]
This is included in pyright 1.1.222, which I just published. It will also be included in the next release of pylance.

**Comment by erictraut** [comment]
It was just pointed out to me that my fix wasn't complete. I fixed the issue with class patterns (the first example above) but not a similar issue with literal patterns (the second example involving enums). I've re-opened the bug and will work on a complete fix.

**Comment by erictraut** [comment]
I've fixed the value pattern logic. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.223, which I just published. It will also be included in the next release of pylance.

**Comment by thisisrandy** [comment]
Thanks as always for the quick response and remediation.
